File: ApiServer/api_server/routers/artifacts.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query, Request

from api_server.schemas import ArtifactListResponse, ArtifactTextResponse
from api_server.services.artifact_service import ArtifactNotFoundError, UnsafeArtifactPathError

router = APIRouter(tags=["artifacts"])


@router.get("/pipelines/{pipeline_id}/stages/{stage_id}/artifacts", response_model=ArtifactListResponse)
def list_stage_artifacts(pipeline_id: str, stage_id: str, request: Request) -> ArtifactListResponse:
    return request.app.state.artifact_service.list_stage_artifacts(pipeline_id, stage_id)


@router.get("/artifacts/file", response_model=ArtifactTextResponse)
def read_artifact_file(request: Request, path: str = Query(...)) -> ArtifactTextResponse:
    try:
        return request.app.state.artifact_service.read_text(path)
    except ArtifactNotFoundError as exc:
        raise HTTPException(status_code=404, detail=f"Artifact not found: {path}") from exc
    except UnsafeArtifactPathError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

File: FrontEnd/src/api/artifacts.ts
import { apiRequest } from "./client";
import type { ArtifactListResponse, ArtifactTextResponse, ReviewAssetsResponse } from "../types/pipeline";

export function listStageArtifacts(pipelineId: string, stageId: string) {
  return apiRequest<ArtifactListResponse>(
    `/api/pipelines/${encodeURIComponent(pipelineId)}/stages/${encodeURIComponent(stageId)}/artifacts`,
  );
}

export function readArtifactFile(path: string) {
  return apiRequest<ArtifactTextResponse>(`/api/artifacts/file?path=${encodeURIComponent(path)}`);
}

export function getStageReviewAssets(pipelineId: string, stageId: string) {
  return apiRequest<ReviewAssetsResponse>(
    `/api/pipelines/${encodeURIComponent(pipelineId)}/stages/${encodeURIComponent(stageId)}/review-assets`,
  );
}
